src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/FormFileModelBinder.cs

```diff
@@ -36,66 +36,151 @@ public Task BindModelAsync(ModelBindingContext bindingContext)
             {
-                var postedFiles = await GetFormFilesAsync(modelName, bindingContext);
-                value = postedFiles.FirstOrDefault();
+                // Create an intermediate list. Will throw it away later.
+                collection = new List<IFormFile>();
             }
-            else if (typeof(IEnumerable<IFormFile>).IsAssignableFrom(bindingContext.ModelType))
+            else if (bindingContext.ModelType == typeof(IFormFile[]))
             {
-                var postedFiles = await GetFormFilesAsync(modelName, bindingContext);
-                value = ModelBindingHelper.ConvertValuesToCollectionType(bindingContext.ModelType, postedFiles);
+                if (bindingContext.ModelMetadata.IsReadOnly)
+                {
+                    // Can't change the length of an existing array or replace it.
+                    return null;
+                }
+
+                // Use a List<IFormFile> for now. Will create an array later.
+                collection = new List<IFormFile>();
             }
             else
             {
-                // This binder does not support the requested type.
-                Debug.Fail("We shouldn't be called without a matching type.");
-                return;
+                if (bindingContext.Model == null)
+                {
+                    if (bindingContext.ModelMetadata.IsReadOnly)
+                    {
+                        // Need a new collection instance but unable to assign it to the property.
+                        return null;
+                    }
+
+                    if (bindingContext.ModelType == typeof(IFormFileCollection))
+                    {
+                        // Special-case creating a custom IFormFileCollection if the property is settable and
+                        // currently null. Use a List<IFormFile> for now. Will create a FileCollection later.
+                        collection = new List<IFormFile>();
+                    }
+                    else
+                    {
+                        // Note this call may return null if the model type cannot be activated.
+                        collection = ModelBindingHelper.CreateCompatibleCollection<IFormFile>(
+                            bindingContext.ModelType,
+                            capacity: null);
+                    }
+                }
+                else
+                {
+                    // Note this cast may fail if the runtime model implements IEnumerable<IFormFile> but not
+                    // ICollection<IFormFile>. Give up in then: Assuming we're not in an odd corner case where the
+                    // property is settable and its declared type is actually assignable from List<IFormFile>.
+                    collection = bindingContext.Model as ICollection<IFormFile>;
+                    collection?.Clear();
+                }
             }
 
-            if (value == null)
+            return collection;
```

Comment:
Is all this stuff really necessary? This just seems so far from simple. At this point you're already using `IFormFile` so you're already coding against our apis. I understand why the code is complicated, just wondering if we need it.